### PR TITLE
feat: add querier interface and initial implementation

### PR DESCRIPTION
**Pull request by srikanthccv**
Part of #1988 

This introduces a querier interface which will be responsible for running the queries for query range requests with optional caching. The current implementation doesn't do any caching for traces and logs query range requests and doesn't support ClickHouse query since it can be very dynamic in nature. Otherwise, the following is the flow.

- Create a cache key for the builder query
- See if there is cached data from the previous run
- Find the missing time range by comparing the current start, and end with the cached start and end
- Query data for the missing time ranges
- Prepare the new merged series from the cache data and query data
- Store this for the potential next

There are some configurable options, 1. Flux interval - querying and caching the more recent data could give incorrect results since some data might not arrive yet. This setting prevents the caching of very recent results that might still be in flux. Default 5 minutes. If you believe this is low, we can bump it up. 2. NoCache - Option for the client to disable the caching entirely.

### CODE REVIEW

File: pkg/query-service/app/querier/querier.go [hunk 1]
```diff
@@ -0,0 +1,429 @@
...
+		err = fmt.Errorf("error in builder queries")
+	}
+	return seriesList, err, errQueriesByName
+}
+
+func (q *querier) runPromQueries(ctx context.Context, params *v3.QueryRangeParamsV3) ([]*v3.Series, error, map[string]string) {
+	seriesList := make([]*v3.Series, 0)
+	errQueriesByName := make(map[string]string)
+	var err error
+	for queryName := range params.CompositeQuery.PromQueries {
```
Comment (nityanandagohain) [inline pkg/query-service/app/querier/querier.go:332]: same here

File: pkg/query-service/app/querier/querier.go [hunk 2]
```diff
@@ -0,0 +1,429 @@
...
+		}
+		if err != nil {
+			errQueriesByName[queryName] = err.Error()
+			continue
+		}
+		misses := q.findMissingTimeRanges(params.Start, params.End, params.Step, cachedData)
+		missedSeries := make([]*v3.Series, 0)
+		cachedSeries := make([]*v3.Series, 0)
+		for _, miss := range misses {
+			query := metricsV3.BuildPromQuery(
```
Comment (nityanandagohain) [inline pkg/query-service/app/querier/querier.go:349]: same we can push these to goruotines

File: pkg/query-service/app/querier/querier_test.go [hunk 3]
```diff
@@ -0,0 +1,511 @@
...
+			},
+		},
+	}
+	q := NewQuerier(opts)
+	expectedTimeRangeInQueryString := []string{
+		fmt.Sprintf("timestamp_ms >= %d AND timestamp_ms <= %d", 1675115596722, 1675115596722+120*60*1000),
+		fmt.Sprintf("timestamp_ms >= %d AND timestamp_ms <= %d", 1675115596722+120*60*1000+1, 1675115596722+180*60*1000),
+	}
+
+	fmt.Println("queries executed", q.QueriesExecuted())
```
Comment (nityanandagohain) [inline pkg/query-service/app/querier/querier_test.go:496]: remove

**Review by nityanandagohain — COMMENTED** [review]
Looks good overall. Moving some of the db calls to goroutines can be taken care in this PR or you can create a new one, will leave it to you.

**Review by nityanandagohain — APPROVED** [review]

**Comment by srikanthccv** [comment]
Merging this, it has been around two weeks.

**Pull request merged**